File: backend/services/service_service.py
from typing import Optional, List, Tuple, Dict, Any
from datetime import datetime, timezone

from db.mongodb import get_database
from models.service import Service, ContentStatus, SEOData, generate_slug
import logging

logger = logging.getLogger(__name__)

class ServiceService:
    COLLECTION = "services"
    
    @staticmethod
    def _serialize_service(service: Service) -> dict:
        """Convert Service model to MongoDB document"""
        doc = service.model_dump()
        doc['created_at'] = doc['created_at'].isoformat() if doc.get('created_at') else None
        doc['updated_at'] = doc['updated_at'].isoformat() if doc.get('updated_at') else None
        doc['archived_at'] = doc['archived_at'].isoformat() if doc.get('archived_at') else None
        doc['status'] = doc['status'].value if hasattr(doc.get('status'), 'value') else doc.get('status')
        doc['price_mode'] = doc['price_mode'].value if hasattr(doc.get('price_mode'), 'value') else doc.get('price_mode')
        if doc.get('seo'):
            doc['seo'] = doc['seo'] if isinstance(doc['seo'], dict) else doc['seo'].model_dump() if hasattr(doc['seo'], 'model_dump') else doc['seo']
        return doc
    
    @staticmethod
    def _deserialize_service(doc: dict) -> dict:
        """Parse MongoDB document dates"""
        for field in ['created_at', 'updated_at', 'archived_at']:
            if doc.get(field) and isinstance(doc[field], str):
                doc[field] = datetime.fromisoformat(doc[field])
        return doc
    
    @staticmethod
    async def create(
        title: str,
        short_description: str,
        cover_image_url: str,
        created_by: str,
        slug: Optional[str] = None,
        **kwargs
    ) -> Optional[Service]:
        """Create a new service"""
        db = get_database()
        
        # Generate slug if not provided
        if not slug:
            slug = generate_slug(title)
        
        # Ensure unique slug
        existing = await db[ServiceService.COLLECTION].find_one({"slug": slug})
        if existing:
            # Append ID suffix
            slug = f"{slug}-{str(uuid.uuid4())[:8]}"
        
        # Handle SEO data
        seo_data = kwargs.pop('seo', None)
        if seo_data and isinstance(seo_data, dict):
            seo_data = SEOData(**seo_data)
        elif not seo_data:
            seo_data = SEOData()
        
        service = Service(
            title=title,
            slug=slug,
            short_description=short_description,
            cover_image_url=cover_image_url,
            created_by=created_by,
            updated_by=created_by,
            seo=seo_data,
            **kwargs
        )
        
        doc = ServiceService._serialize_service(service)
        await db[ServiceService.COLLECTION].insert_one(doc)
        return service
    
    @staticmethod
    async def get_by_id(service_id: str) -> Optional[Service]:
        """Get service by ID"""
        db = get_database()
        doc = await db[ServiceService.COLLECTION].find_one({"id": service_id}, {"_id": 0})
        if not doc:
            return None
        doc = ServiceService._deserialize_service(doc)
        return Service(**doc)
    
    @staticmethod
    async def get_by_slug(slug: str) -> Optional[Service]:
        """Get service by slug"""
        db = get_database()
        doc = await db[ServiceService.COLLECTION].find_one({"slug": slug}, {"_id": 0})
        if not doc:
            return None
        doc = ServiceService._deserialize_service(doc)
        return Service(**doc)
    
    @staticmethod
    async def get_list(
        page: int = 1,
        page_size: int = 20,
        status: Optional[ContentStatus] = None,
        archived: bool = False,
        search: Optional[str] = None
    ) -> Tuple[List[Service], int]:
        """Get paginated list of services for admin"""
        db = get_database()
        
        query: Dict[str, Any] = {}
        
        # Filter by archived status
        if archived:
            query["archived_at"] = {"$ne": None}
        else:
            query["archived_at"] = None
        
        # Filter by status
        if status:
            query["status"] = status.value
        
        # Search
        if search:
            query["$or"] = [
                {"title": {"$regex": search, "$options": "i"}},
                {"short_description": {"$regex": search, "$options": "i"}}
            ]
        
        total = await db[ServiceService.COLLECTION].count_documents(query)
        
        skip = (page - 1) * page_size
        cursor = db[ServiceService.COLLECTION].find(query, {"_id": 0}).skip(skip).limit(page_size).sort("sort_order", 1)
        docs = await cursor.to_list(page_size)
        
        services = []
        for doc in docs:
            doc = ServiceService._deserialize_service(doc)
            services.append(Service(**doc))
        
        return services, total
    
    @staticmethod
    async def get_published() -> List[Service]:
        """Get all published, non-archived services for public"""
        db = get_database()
        
        query = {
            "status": ContentStatus.PUBLISHED.value,
            "archived_at": None
        }
        
        cursor = db[ServiceService.COLLECTION].find(query, {"_id": 0}).sort("sort_order", 1)
        docs = await cursor.to_list(100)
        
        services = []
        for doc in docs:
            doc = ServiceService._deserialize_service(doc)
            services.append(Service(**doc))
        
        return services
    
    @staticmethod
    async def get_published_by_slug(slug: str) -> Optional[Service]:
        """Get published service by slug for public"""
        db = get_database()
        
        query = {
            "slug": slug,
            "status": ContentStatus.PUBLISHED.value,
            "archived_at": None
        }
        
        doc = await db[ServiceService.COLLECTION].find_one(query, {"_id": 0})
        if not doc:
            return None
        doc = ServiceService._deserialize_service(doc)
        return Service(**doc)
    
    @staticmethod
    async def update(
        service_id: str,
        updated_by: str,
        **kwargs
    ) -> Optional[Service]:
        """Update service fields"""
        db = get_database()
        
        update_data = {
            "updated_at": datetime.now(timezone.utc).isoformat(),
            "updated_by": updated_by
        }
        
        for key, value in kwargs.items():
            if value is not None:
                if key == 'status' and hasattr(value, 'value'):
                    update_data[key] = value.value
                elif key == 'price_mode' and hasattr(value, 'value'):
                    update_data[key] = value.value
                elif key == 'seo' and hasattr(value, 'model_dump'):
                    update_data[key] = value.model_dump()
                elif key == 'seo' and isinstance(value, dict):
                    update_data[key] = value
                else:
                    update_data[key] = value
        
        # Ensure unique slug if updating
        if 'slug' in update_data:
            existing = await db[ServiceService.COLLECTION].find_one({
                "slug": update_data['slug'],
                "id": {"$ne": service_id}
            })
            if existing:
                return None  # Slug conflict
        
        result = await db[ServiceService.COLLECTION].update_one(
            {"id": service_id},
            {"$set": update_data}
        )
        
        if result.modified_count == 0:
            return None
        
        return await ServiceService.get_by_id(service_id)
    
    @staticmethod
    async def update_status(service_id: str, status: ContentStatus, updated_by: str) -> Optional[Service]:
        """Update service status (publish/unpublish)"""
        return await ServiceService.update(service_id, updated_by, status=status)
    
    @staticmethod
    async def archive(service_id: str, updated_by: str) -> Optional[Service]:
        """Soft delete (archive) a service"""
        db = get_database()
        
        result = await db[ServiceService.COLLECTION].update_one(
            {"id": service_id},
            {"$set": {
                "archived_at": datetime.now(timezone.utc).isoformat(),
                "updated_at": datetime.now(timezone.utc).isoformat(),
                "updated_by": updated_by
            }}
        )
        
        if result.modified_count == 0:
            return None
        
        return await ServiceService.get_by_id(service_id)
    
    @staticmethod
    async def restore(service_id: str, updated_by: str) -> Optional[Service]:
        """Restore an archived service"""
        db = get_database()
        
        result = await db[ServiceService.COLLECTION].update_one(
            {"id": service_id},
            {"$set": {
                "archived_at": None,
                "updated_at": datetime.now(timezone.utc).isoformat(),
                "updated_by": updated_by
            }}
        )
        
        if result.modified_count == 0:
            return None
        
        return await ServiceService.get_by_id(service_id)
    
    @staticmethod
    async def reorder(items: List[dict], updated_by: str) -> bool:
        """Bulk reorder services"""
        db = get_database()
        
        for item in items:
            await db[ServiceService.COLLECTION].update_one(
                {"id": item["id"]},
                {"$set": {
                    "sort_order": item["sort_order"],
                    "updated_at": datetime.now(timezone.utc).isoformat(),
                    "updated_by": updated_by
                }}
            )
        
        return True

import uuid

File: src/lib/services/service-service.ts
import { v4 as uuidv4 } from 'uuid'
import { getCollection, COLLECTIONS } from '@/lib/db/mongodb'
import type { Service, ServiceStatus, PriceMode, ServiceSEO } from '@/lib/models/types'

function generateSlug(title: string): string {
  return title
    .toLowerCase()
    .replace(/[^\w\sğüşıöç-]/g, '')
    .replace(/\s+/g, '-')
    .replace(/ğ/g, 'g')
    .replace(/ü/g, 'u')
    .replace(/ş/g, 's')
    .replace(/ı/g, 'i')
    .replace(/ö/g, 'o')
    .replace(/ç/g, 'c')
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '')
}

export class ServiceService {
  // Get all services (admin)
  static async list(
    page: number = 1,
    limit: number = 20,
    filters?: { status?: ServiceStatus; archived?: boolean; search?: string }
  ): Promise<{ services: Service[]; total: number }> {
    const servicesCollection = await getCollection<Service>(COLLECTIONS.SERVICES)
    
    const filter: Record<string, unknown> = {}
    
    if (filters?.archived) {
      filter.archived_at = { $ne: null }
    } else {
      filter.archived_at = null
    }
    
    if (filters?.status) {
      filter.status = filters.status
    }
    
    if (filters?.search) {
      filter.$or = [
        { title: { $regex: filters.search, $options: 'i' } },
        { short_description: { $regex: filters.search, $options: 'i' } }
      ]
    }
    
    const total = await servicesCollection.countDocuments(filter)
    const services = await servicesCollection
      .find(filter)
      .sort({ sort_order: 1, created_at: -1 })
      .skip((page - 1) * limit)
      .limit(limit)
      .toArray()
    
    return { services, total }
  }
  
  // Get published services (public)
  static async getPublished(): Promise<Service[]> {
    const servicesCollection = await getCollection<Service>(COLLECTIONS.SERVICES)
    
    return servicesCollection
      .find({
        status: 'published',
        archived_at: null
      })
      .sort({ sort_order: 1 })
      .toArray()
  }
  
  // Get by ID
  static async getById(id: string): Promise<Service | null> {
    const servicesCollection = await getCollection<Service>(COLLECTIONS.SERVICES)
    return servicesCollection.findOne({ id })
  }
  
  // Get by slug (public)
  static async getPublishedBySlug(slug: string): Promise<Service | null> {
    const servicesCollection = await getCollection<Service>(COLLECTIONS.SERVICES)
    return servicesCollection.findOne({
      slug,
      status: 'published',
      archived_at: null
    })
  }
  
  // Create service
  static async create(data: {
    title: string
    short_description: string
    long_description?: string
    cover_image_url: string
    cover_image_alt?: string
    icon?: string
    price_mode?: PriceMode
    price_value?: number
    tags?: string[]
    category?: string
    seo?: ServiceSEO
    status?: ServiceStatus
    created_by?: string
  }): Promise<Service> {
    const servicesCollection = await getCollection<Service>(COLLECTIONS.SERVICES)
    
    // Generate unique slug
    let slug = generateSlug(data.title)
    let counter = 1
    while (await servicesCollection.findOne({ slug })) {
      slug = `${generateSlug(data.title)}-${counter}`
      counter++
    }
    
    // Get max sort order
    const maxSort = await servicesCollection
      .find({})
      .sort({ sort_order: -1 })
      .limit(1)
      .toArray()
    const sortOrder = maxSort.length > 0 ? maxSort[0].sort_order + 1 : 0
    
    const now = new Date()
    const service: Service = {
      id: uuidv4(),
      title: data.title,
      slug,
      short_description: data.short_description,
      long_description: data.long_description || null,
      cover_image_url: data.cover_image_url,
      cover_image_alt: data.cover_image_alt || null,
      icon: data.icon || null,
      price_mode: data.price_mode || 'contact',
      price_value: data.price_value || null,
      tags: data.tags || [],
      category: data.category || null,
      seo: data.seo || { meta_title: null, meta_description: null, og_image_url: null },
      status: data.status || 'draft',
      sort_order: sortOrder,
      archived_at: null,
      created_at: now,
      updated_at: now,
      created_by: data.created_by || null,
      updated_by: null
    }
    
    await servicesCollection.insertOne(service)
    return service
  }
  
  // Update service
  static async update(id: string, data: Partial<Service>, updatedBy?: string): Promise<Service | null> {
    const servicesCollection = await getCollection<Service>(COLLECTIONS.SERVICES)
    
    const updateData: Record<string, unknown> = {
      ...data,
      updated_at: new Date(),
      updated_by: updatedBy || null
    }
    
    // Don't allow updating certain fields
    delete updateData.id
    delete updateData.created_at
    delete updateData.created_by
    
    await servicesCollection.updateOne(
      { id },
      { $set: updateData }
    )
    
    return this.getById(id)
  }
  
  // Update status
  static async updateStatus(id: string, status: ServiceStatus): Promise<Service | null> {
    return this.update(id, { status })
  }
  
  // Delete (archive)
  static async archive(id: string): Promise<boolean> {
    const servicesCollection = await getCollection<Service>(COLLECTIONS.SERVICES)
    
    const result = await servicesCollection.updateOne(
      { id },
      { $set: { archived_at: new Date(), status: 'archived' } }
    )
    
    return result.modifiedCount > 0
  }
  
  // Restore
  static async restore(id: string): Promise<Service | null> {
    const servicesCollection = await getCollection<Service>(COLLECTIONS.SERVICES)
    
    await servicesCollection.updateOne(
      { id },
      { $set: { archived_at: null, status: 'draft' } }
    )
    
    return this.getById(id)
  }
  
  // Reorder
  static async reorder(items: Array<{ id: string; sort_order: number }>): Promise<void> {
    const servicesCollection = await getCollection<Service>(COLLECTIONS.SERVICES)
    
    const bulkOps = items.map(item => ({
      updateOne: {
        filter: { id: item.id },
        update: { $set: { sort_order: item.sort_order, updated_at: new Date() } }
      }
    }))
    
    await servicesCollection.bulkWrite(bulkOps)
  }
}
